File: 3621.number-of-integers-with-popcount-depth-equal-to-k-i.py
#
# @lc app=leetcode id=3621 lang=python3
#
# [3621] Number of Integers With Popcount-Depth Equal to K I
#

# @lc code=start
class Solution:
    def popcountDepth(self, n: int, k: int) -> int:
        from functools import lru_cache
        # Step 1: Precompute popcount-depth for possible s up to sufficient max_bits
        max_bits = n.bit_length() + 2  # Sufficiently covers all possible set bit counts
        depth = [-1] * (max_bits + 1)
        def get_depth(x):
            if x == 1:
                return 0
            if depth[x] != -1:
                return depth[x]
            d = 1 + get_depth(bin(x).count('1'))
            depth[x] = d
            return d
        for s in range(1, max_bits + 1):
            get_depth(s)
        # Step 2: Collect valid s for which popcount-depth(s) == k-1
        valid_s = []
        for s in range(1, max_bits + 1):
            if get_depth(s) == k - 1:
                valid_s.append(s)
        # Step 3: Count numbers <= n with exactly s set bits
        def count_numbers(n, s):
            bits = []
            tmp = n
            while tmp:
                bits.append(tmp & 1)
                tmp >>= 1
            bits = bits[::-1]
            m = len(bits)
            @lru_cache(None)
            def dp(pos, cnt, tight):
                if cnt < 0:
                    return 0
                if pos == m:
                    return 1 if cnt == 0 else 0
                res = 0
                up = bits[pos] if tight else 1
                for b in range(up + 1):
                    res += dp(pos + 1, cnt - b, tight and (b == up))
                return res
            return dp(0, s, True)
        # Step 5: Handle special case k == 0
        if k == 0:
            return 1 if n >= 1 else 0
        # Step 4: Sum results for all valid s and verify full coverage
        ans = 0
        for s in valid_s:
            ans += count_numbers(n, s)
        # Exclude x=0 (not in [1, n]) if needed
        if k == 1:
            ans -= 1
        # Step 6: Final verification is implicit in code structure and coverage
        return ans
# @lc code=end

File: 3621.number-of-integers-with-popcount-depth-equal-to-k-i.java
#
# @lc app=leetcode id=3621 lang=java
#
# [3621] Number of Integers With Popcount-Depth Equal to K I
#

# @lc code=start
class Solution {
    // Memoization for popcount-depth
    private int[] depth = new int[65];
    private boolean[] computed = new boolean[65];
    
    private int popcount(int x) {
        int res = 0;
        while (x > 0) {
            res += x & 1;
            x >>= 1;
        }
        return res;
    }
    
    private int getDepth(int x) {
        if (x < 1) return -1;
        if (x == 1) return 0;
        if (computed[x]) return depth[x];
        depth[x] = 1 + getDepth(popcount(x));
        computed[x] = true;
        return depth[x];
    }
    
    // Digit DP: count numbers <= n with exactly k set bits
    private long countWithKSetBits(long n, int k) {
        int[] bits = new int[65];
        int len = 0;
        long tmp = n;
        while (tmp > 0) {
            bits[len++] = (int)(tmp & 1);
            tmp >>= 1;
        }
        // dp[pos][cnt][tight], tight: 1 if prefix matches n, 0 if already less
        Long[][][] dp = new Long[len + 1][k + 2][2];
        return dfs(bits, len - 1, k, 1, dp);
    }
    
    private long dfs(int[] bits, int pos, int k, int tight, Long[][][] dp) {
        if (k < 0) return 0;
        if (pos < 0) return k == 0 ? 1 : 0;
        if (dp[pos][k][tight] != null) return dp[pos][k][tight];
        long res = 0;
        int limit = tight == 1 ? bits[pos] : 1;
        for (int d = 0; d <= limit; ++d) {
            int newTight = (tight == 1 && d == limit) ? 1 : 0;
            res += dfs(bits, pos - 1, k - d, newTight, dp);
        }
        dp[pos][k][tight] = res;
        return res;
    }
    
    public long popcountDepth(long n, int k) {
        if (k == 0) return 0;
        computed[1] = true; depth[1] = 0;
        for (int s = 1; s <= 60; ++s) getDepth(s);
        long ans = 0;
        for (int s = 1; s <= 60; ++s) {
            if (getDepth(s) == k) {
                long cnt = countWithKSetBits(n, s);
                // Exclude 0 if counted (since [1, n])
                if (s == 1 && k == 1) cnt--;
                ans += cnt;
            }
        }
        return ans;
    }
}
# @lc code=end